src/components/Api.js

```diff
@@ -0,0 +1,94 @@
+export default class Api {
+    constructor(options) {
+        this._options = options;
+        this._baseUrl = this._options.baseUrl;
+        this.isComplete = false;
+    }
+
+    //the basic root functions that send requests out or brings them in- parameterized with ammendments to the fetch body
+
+    _fetchX = (endpoint, how) => {
+        this._options.method = how;
+        return fetch(this._baseUrl + endpoint, this._options)
+        .then(res => {
+            if (res.ok) {
+                return res.json();
+            }
+            return Promise.reject(`Error: ${res.status}`);
+        })
+        .then(data => {
+            this.isComplete = true;
+            return data;
+        })
+        .catch((err) => {
+            console.log(`WHOOPSIES ON THE FETCH: the ` + how + ` has failed in execution, reporting ` + err);
+        })
+        .finally(() => {
+            this._options.method = "";
+            this.isComplete = false;
+        });
+    }
+
+    _sendX = (endpoint, parcel, how) => {
+        this._options.method = how;
+        this._options.body = JSON.stringify(parcel);   
+        return fetch(this._baseUrl + endpoint, this._options)
+        .then(res => {
+            if (res.ok) {
+                console.log(res.ok);
+                console.log(res.status);
+                return res.json();
+            }
+            return Promise.reject(`uh ohs on the send: ${res.status}`);
```

Comment:
```js
.then(res => {
                if (res.ok) {
                    return res.json();
                }
                return Promise.reject(`Error ${res.status}`);
            })

```

The code of `ok` checking is repeated in **every** request of the project. To avoid it you need to make a special method `_checkResponse` that will check any response from the server if it’s ok or not.


```js
_checkResponse(res) {
        // here is the code of the checking
    }

```

And now you can replace redundant repeating with the only line:


```js
 .then(this._checkResponse)

```

Please, pay attention that you need to pass only the **reference** for the method rather than the call. 

**Could be improved**

You can make a special method (function ) for fetching and checking responses **not to duplicate** it in every request:

```js

  _request(url, options) {
    return fetch(url, options).then(this._checkResponse)
  }

```

And now you can safely replace `fetch` with `this._request` and remove repeated checking. And you don’t need to change anything else.